src/goto-instrument/contracts/contracts.cpp

```diff
@@ -450,44 +514,60 @@ void code_contractst::check_apply_loop_contracts(
+        new_decreases_vars[i], decreases_clause_exprs[i]};
+      new_decreases_assignment.add_source_location() = loop_head_location;
+      converter.goto_convert(
+        new_decreases_assignment, pre_loop_end_instrs, mode);
     }
 
     // Generate: assertion that the multidimensional decreases clause's value
-    // after the loop is smaller than the value before the loop.
-    // Here, we use the lexicographic order.
+    // after the loop is lexicographically smaller than its initial value.
     code_assertt monotonic_decreasing_assertion{
       generate_lexicographic_less_than_check(
         new_decreases_vars, old_decreases_vars)};
-    monotonic_decreasing_assertion.add_source_location() =
-      loop_head->source_location();
+    monotonic_decreasing_assertion.add_source_location() = loop_head_location;
     converter.goto_convert(
-      monotonic_decreasing_assertion, generated_code, mode);
-    generated_code.instructions.back().source_location_nonconst().set_comment(
-      "Check decreases clause on loop iteration");
+      monotonic_decreasing_assertion, pre_loop_end_instrs, mode);
+    pre_loop_end_instrs.instructions.back()
+      .source_location_nonconst()
+      .set_comment("Check decreases clause on loop iteration");
 
     // Discard the temporary variables that store decreases clause's value
     for(size_t i = 0; i < old_decreases_vars.size(); i++)
     {
-      generated_code.add(goto_programt::make_dead(
-        old_decreases_vars[i], loop_head->source_location()));
-      generated_code.add(goto_programt::make_dead(
-        new_decreases_vars[i], loop_head->source_location()));
+      pre_loop_end_instrs.add(
+        goto_programt::make_dead(old_decreases_vars[i], loop_head_location));
+      pre_loop_end_instrs.add(
+        goto_programt::make_dead(new_decreases_vars[i], loop_head_location));
     }
   }
 
   insert_before_swap_and_advance(
     goto_function.body,
     loop_end,
-    add_pragma_disable_assigns_check(generated_code));
+    add_pragma_disable_assigns_check(pre_loop_end_instrs));
 
   // change the back edge into assume(false) or assume(guard)
   loop_end->turn_into_assume();
   loop_end->set_condition(boolean_negate(loop_end->get_condition()));
+
+  // Instructions to make all the temporaries go dead.
+  // goto_programt pre_loop_exit_instrs;
+  // pre_loop_exit_instrs.add(
+  //   goto_programt::make_dead(in_base_case, loop_head_location));
+  // pre_loop_exit_instrs.add(
+  //   goto_programt::make_dead(initial_invariant_val, loop_head_location));
+  // for(const auto &v : history_var_map)
+  // {
+  //   pre_loop_exit_instrs.add(goto_programt::make_dead(
+  //     static_cast<const symbol_exprt &>(v.first), loop_head_location));
+  // }
+  //
+  // TODO: ^^^ I think these should be inserted at all exit points:
+  // falsification of guard, break statements, or even `goto` jumps
+  // out of a loop.
+  // Is there an easy way to add these everywhere?
```

Comment:
at any point where the jump target or the next instruction is not in the `loop` set ?

Reply:
Thanks. Implemented it as you suggested.